src/edu/washington/escience/myria/operator/CSVFileScanFragment.java

```diff
@@ -181,142 +187,184 @@ protected TupleBatch fetchNextReady() throws IOException, DbException {
-                  source.getInputStream(partitionEndByteRange + 1, movingEndByte);
-              partitionInputStream =
-                  new SequenceInputStream(beginningOfRecord, concatenateEndOfRecord);
-              parser =
-                  new CSVParser(
-                      new BufferedReader(new InputStreamReader(partitionInputStream)),
-                      CSVFormat.newFormat(delimiter).withQuote(quote).withEscape(escape),
-                      bytePositionAtBeginningOfRecord,
-                      0);
-              iterator = parser.iterator();
-              onLastRow = true;
-              record = iterator.next();
+
+      if (record != null) {
+        /*
+         * Here, we read the first byte of the stream if this is not the first worker. This is simply an optimization.
+         * If the previous character is not a new line, we discard the line.
+         */
+        if (lineNumber - 1 == 0 && partitionStartByteRange != 0) {
+          char currentChar = (char) partitionInputStream.read();
+          if (currentChar != '\n' && currentChar != '\r') {
+            discardedRecord = true;
+            /*
+             * This handles closes the stream in corner cases in which a worker might only have a fragment of a tuple as
+             * its entire partition. Since the server makes sure to provide workers a partition with a minimum size,
+             * this case could only happen with very large rows.
+             */
+            if (onLastRow) {
+              parser.close();
               break;
             }
-            dataChar = startStreamReader.read();
+          } else {
+            /*
+             * This handles the case in which the splitting occurs in the middle of \r\n.
+             */
+            currentChar = (char) partitionInputStream.read();
+            if (currentChar == '\n') {
+              discardedRecord = true;
+            } else {
+              partitionInputStream =
+                  source.getInputStream(adjustedStartByteRange, adjustedEndByteRange);
+            }
           }
-          startStreamReader.close();
         }
-      }
-      // This is a partial fragment, columns are missing and we need to read more bytes
-      if (record.size() < schema.numColumns()) {
-        if (lineNumber - 1 != 0 && !isLastWorker) {
-          onLastRow = true;
-          long bytePositionAtBeginningOfRecord = record.getCharacterPosition();
-          if (!fixingStartByte) {
-            fixingStartByte = true;
-            partitionStartByteRange += bytePositionAtBeginningOfRecord;
+
+        /*
+         * Here, if we are on the last row, we make sure to read the entire row until we either hit a new line or until
+         * we have read the entire file (again, this is for the case where a single worker might be reading a single
+         * large row that was split among other workers)
+         */
+        if (record.size() == schema.numColumns() && onLastRow && !isLastWorker) {
```

Comment:
Actually this is a little trickier than I thought. I guess your assumption is that any trailing record fragment will be parsed as an incomplete record (`record.size() < schema.numColumns()`), in which case we'll just fetch more data, scanning for the first newline to mark the end of the trailing record.  But as I said, imagine an example where each record field is a quoted string, and the trailing fragment is truncated in the middle of the first field, so the parser can't parse even one field of the trailing record. My assumption is that `iterator.hasNext()` would return `false` after the preceding record, so this fragment would not be marked as the last record. Therefore we would miss this record in its entirety, since the next partition would discard it. So how do we handle this? This should be pretty simple to test.